nsxt/segment_port_common.go

```diff
@@ -0,0 +1,692 @@
+		obj.Description = &description
+	}
+
+	err := nsxtPolicySegmentPortProfilesSetInStruct(d, &obj)
+	if err != nil {
+		return model.Infra{}, err
+	}
+
+	childSegmentPort := model.ChildSegmentPort{
+		SegmentPort:     &obj,
+		ResourceType:    "ChildSegmentPort",
+		MarkedForDelete: &isDestroy,
+	}
+
+	// Segment
+	child, err := vAPIConversion(childSegmentPort, model.ChildSegmentPortBindingType())
+	if err != nil {
+		return model.Infra{}, fmt.Errorf("Error handling the SegmentPort hierarchial API construction : %v", err)
+	}
+	segmentChildren := []*data.StructValue{child}
+	segmentId := getSegmentIdFromSegPath(segmentPath)
+	segmentTargetType := "Segment"
+	childSegment := model.ChildResourceReference{
+		Id:           &segmentId,
+		ResourceType: "ChildResourceReference",
+		TargetType:   &segmentTargetType,
+		Children:     segmentChildren,
+	}
+
+	// Tier1
+	child, err = vAPIConversion(childSegment, model.ChildResourceReferenceBindingType())
+	if err != nil {
+		return model.Infra{}, fmt.Errorf("Error handling the Tier1 gw hierarchial API construction : %v", err)
+	}
+	if isT1Segment(segmentPath) {
+		t1Children := []*data.StructValue{child}
+		t1GwId := getT1IdFromSegPath(segmentPath)
+		tier1TargetType := "Tier1"
+		childTier1Gw := model.ChildResourceReference{
+			Id:           &t1GwId,
+			ResourceType: "ChildResourceReference",
+			TargetType:   &tier1TargetType,
+			Children:     t1Children,
+		}
+
+		child, err = vAPIConversion(childTier1Gw, model.ChildResourceReferenceBindingType())
+		if err != nil {
+			return model.Infra{}, fmt.Errorf("Error handling the Infra hierarchial API construction : %v", err)
+		}
+	}
+	// Infra
+	infraType := "Infra"
+	infraStruct := model.Infra{
+		Children:     []*data.StructValue{child},
+		ResourceType: &infraType,
+	}
+
+	return infraStruct, nil
+}
+
+func vAPIConversion(golangValue interface{}, bindingType bindings.BindingType) (*data.StructValue, error) {
+	converter := bindings.NewTypeConverter()
+	dataValue, errors := converter.ConvertToVapi(golangValue, bindingType)
+	if errors != nil {
+		return &data.StructValue{}, fmt.Errorf("Error converting Segment Child: %v", errors[0])
+	}
+	return dataValue.(*data.StructValue), nil
+}
+
+func nsxtPolicySegmentPortProfilesSetInStruct(d *schema.ResourceData, obj *model.SegmentPort) error {
+	children := []*data.StructValue{}
+
+	child, err := nsxtPolicyPortDiscoveryProfileSetInStruct(d)
+	if err != nil {
+		return err
+	}
+
+	if child != nil {
+		children = append(children, child)
+	}
+
+	child, err = nsxtPolicyPortQosProfileSetInStruct(d)
+	if err != nil {
+		return err
+	}
+
+	if child != nil {
+		children = append(children, child)
+	}
+
+	child, err = nsxtPolicyPortSecurityProfileSetInStruct(d)
+	if err != nil {
+		return err
+	}
+
+	if child != nil {
+		children = append(children, child)
+	}
+
+	obj.Children = children
+	return nil
+
+}
+
+func nsxtPolicyPortDiscoveryProfileSetInStruct(d *schema.ResourceData) (*data.StructValue, error) {
```

Comment:
Question: how feasible would be to generalize the logic in segment_common.go so that we can use it for setting profiles either in ports or segments?